### PR TITLE
Allow .parse to load from string

### PR DESCRIPTION
**Pull request by kmcphillips**
I'm doing some work of loading weights from files, and I want to be able to have values in an input file:

```
weight: 10.5 kg
```

And then create a weight object out of it:

```
Measured::Weight.parse("10.5 kg")
```

It's way slower than taking in the number and the symbol, but we'd be parsing those strings from the file first anyway.

### CODE REVIEW

File: lib/measured/measurable.rb [hunk 1]
```diff
@@ -1,6 +1,8 @@
 class Measured::Measurable < Numeric
   include Measured::Arithmetic
 
+  PARSE_REGEX = /\A\s*([0-9.]+)\s*([a-zA-Z_-]+)\s*\Z/
```
Comment (kmcphillips) [inline lib/measured/measurable.rb]: Accept only one line, any blank characters, then any digits with an optional dot, then optional blank characters, than any letters numbers dashes underscores, then any blank characters.

Comment (thegedge) [inline lib/measured/measurable.rb]: `Unit` does [this](https://github.com/Shopify/measured/blob/2a05bfc90ea95a29fc11cd693b20c9e84eec2be7/lib/measured/unit.rb#L66-L71), for conversion values when building a unit system, so we should probably share the parsing code between them.

Two cases missing atm: negative numbers (e.g., -1°C) and unit names with spaces (e.g., nautical mile).

Comment (jonathankwok) [inline lib/measured/measurable.rb]: So the fact that we need a review comment means we should do something about the regex clarity. Free-spacing regex is pretty fantastic, by using `/x` which ignores whitespaces.

```ruby
# before
/\A\s*([0-9.]+)\s*([a-zA-Z_-]+)\s*\Z/

# after
/
  \A                # start of line
  \s*               # any number of blank characters
  ([0-9.]+)         # one or more digits or decimals
  \s*               # any number of blank characters
  ([a-zA-Z_-]+)     # one or more letters, dashes, or underscores
  \s*               # any number of blank characters
  \Z                # end of line
/x
```

Comment (jonathankwok) [inline lib/measured/measurable.rb]: Also in doing that I see that we accept multiple dots. Does that make sense for measured? Numbers/quantities can't have multiple dots.

```ruby
irb(main):001:0> pattern = /
irb(main):002:0/   \A                # start of line
irb(main):003:0/   \s*               # any number of blank characters
irb(main):004:0/   ([0-9.]+)         # one or more digits or decimals
irb(main):005:0/   \s*               # any number of blank characters
irb(main):006:0/   ([a-zA-Z_-]+)     # one or more letters, dashes, or underscores
irb(main):007:0/   \s*               # any number of blank characters
irb(main):008:0/   \Z                # end of line
irb(main):009:0/ /x
=> /
  \A                # start of line
  \s*               # any number of blank characters
  ([0-9.]+)         # one or more digits or decimals
  \s*               # any number of blank characters
  ([a-zA-Z_-]+)     # one or more letters, dashes, or underscores
  \s*               # any number of blank characters
  \Z                # end of line
/x
irb(main):010:0> pattern =~ "1.21.32 gigawatts"
=> 0
```

Comment (jonathankwok) [inline lib/measured/measurable.rb]: Finally, on that topic - we should consider commas and periods inter-changable. [A non-trivial amount of countries in the world represent the idea of `1.21` as `1,21`.](http://brilliantmaps.com/decimals/)

**Comment by kmcphillips** [comment]
Fantastic answers!

Splitting out the regex for readability is a wonderful idea. I'll do that, and deal with negatives and single period. I'll also test those cases.

I'll see about refactoring the `Unit` class to share this logic. I'm sure we can use a single parser, or at least a single approach. I really don't want to just do `.strip.split(" ", 2)` because I want to support `12cm` as that's easy to parse but has no spaces.

File: lib/measured/measurable.rb [hunk 2]
```diff
@@ -1,6 +1,26 @@
...
   include Measured::Arithmetic
 
+  PARSE_REGEX = /
+    \A         # beginning of input
+    \s*        # optionally any whitespace
+    (          # capture the value
+      -?       # number can be negative
+      \d+      # must have some digits
+      (?:      # do not capture
+        \.     # period to split fractional part
```
Comment (jonathankwok) [inline lib/measured/measurable.rb]: We should permit the use of commas in place of periods. (https://github.com/Shopify/measured/pull/81#discussion_r102552174)

File: README.md [hunk 3]
```diff
@@ -56,6 +56,12 @@ rescue Measured::UnitError
 end
 ```
 
+Parse from string without having to split out the value and unit first:
+
+```ruby
+Measured::Weight.parse("123 grams")
```
Comment (thegedge) [inline README.md]: nit: to be aligned with the other comments, show the `#inspect` after. Maybe also show that you don't need a space.

File: lib/measured/measurable.rb [hunk 4]
```diff
@@ -1,6 +1,26 @@
...
+      )?       # fractional part is optional
+    )
+    \s*        # optionally any space between number and unit
+    (          # capture the unit
+      [\w\s-]  # any word or space or dash to
+      +?       # capture it lazily to not take trailing whitespace
+    )
+    \s*        # optionally any whitespace
+    \Z         # end of unit
+  /x
```
Comment (thegedge) [inline lib/measured/measurable.rb]: We can iterate on it, but some other things I'd love to be able to support:

- `1/3 grams`
- `5e-10 cm`
- `.5 metres`

File: lib/measured/measurable.rb [hunk 5]
```diff
@@ -1,6 +1,26 @@
...
+        \d+    # some digits after it
+      )?       # fractional part is optional
+    )
+    \s*        # optionally any space between number and unit
+    (          # capture the unit
+      [\w\s-]  # any word or space or dash to
+      +?       # capture it lazily to not take trailing whitespace
+    )
+    \s*        # optionally any whitespace
+    \Z         # end of unit
```
Comment (thegedge) [inline lib/measured/measurable.rb]: nit: "end of input"

File: test/measurable_test.rb [hunk 6]
```diff
@@ -94,7 +94,7 @@ class Measured::MeasurableTest < ActiveSupport::TestCase
 
   test ".unit_names_with_aliases returns all units" do
     assert_equal(
-      %w(arcane fire fireball fireballs ice magic_missile magic_missiles ultima),
+      %w(arcane fire fireball fireballs ice magic\ missile magic_missile magic_missiles ultima),
```
Comment (thegedge) [inline test/measurable_test.rb]: 😭 

But 👌  for having this case tested 😄

File: lib/measured/measurable.rb (hunk 2 above)
Comment (thegedge) [inline lib/measured/measurable.rb]: I think commas are pervasive enough to support, but a larger question to ask is how far do we go with localization here. Can take that to a :octocat: issue.

File: README.md (hunk 3 above)
Comment (kmcphillips) [inline README.md]: We don't show the result for all operations. And this is just showing the parser syntax.

Comment (kmcphillips) [inline README.md]: But I think you're right anyway. 😄 

Added more detail.

Comment (thegedge) [inline README.md]: Up to you :)

**Comment by kmcphillips** [comment]
Ok, reping.

Since we parse in both the unit definition and now on instantiating a measurement, I extracted a `Measured::Parser` class. It's nice because it now encapsulates the regex. Split it up as suggested, with comments. Way built out what's in that regex. Added a ton of edge cases. Shared the parsing.

Some tests are duplicated, but I'm ok with that. Had to add a case for a unit with a space in it.

If it's too much I can split the PR up. But I think it's easy enough to follow.

**Comment by kmcphillips** [comment]
Oh and @jonathankwok I went back and forth on the `,` parsing and decided _not_ to include it. I would be doing a `value.gsub(",", ".")` since the numeric parsing in ruby truncates after the comma. It doesn't align with the language, so we won't do it here.

File: lib/measured/unit.rb [hunk 7]
```diff
@@ -64,9 +64,14 @@ def conversion_string
   end
 
   def parse_value(tokens)
-    tokens = tokens.split(" ", 2) if tokens.is_a?(String)
-
-    raise Measured::UnitError, "Cannot parse 'number unit' or [number, unit] formatted tokens from #{tokens}." unless tokens.size == 2
```
Comment (kmcphillips) [inline lib/measured/unit.rb]: This is all now much more robust. Error messaging is cleaner, and it allows units to be declared with more edge cases in mind.

File: lib/measured/parser.rb [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+      -?       # number can be negative
+      \d+      # must have some digits
+      (?:      # do not capture
+        [\.\/] # period or slash to split fractional part
+        \d+    # some digits after it
+      )?       # fractional part is optional
+    )
+    \s*        # optionally any space between number and unit
+    (          # capture the unit
+      [a-zA-Z] #
```
Comment (jonathankwok) [inline lib/measured/parser.rb]: This is because we want the unit to start with a letter, right? Maybe a comment to suggest as much?

File: lib/measured/parser.rb [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+      [\w-]*   # any word characters or dashes
+      (?:
```
Comment (jonathankwok) [inline lib/measured/parser.rb]: Can you add comments to explain what this non-capturing group does?

**Comment by kmcphillips** [comment]
@jonathankwok much clearer regex comments now

File: lib/measured/parser.rb [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+      -?       # number can be negative
+      \d+      # must have some digits
+      (?:      # do not capture
+        [\.\/] # period or slash to split fractional part
+        \d+    # some digits after it
+      )?       # fractional part is optional
+    )
+    \s*        # optionally any space between number and unit
+    (          # capture the unit
+      [a-zA-Z] # unit maust start with a letter
```
Comment (jonathankwok) [inline lib/measured/parser.rb]: `maust`

File: lib/measured/parser.rb [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+      [\w-]*   # any word characters or dashes
+      (?:      # non capturing group
+        \s+    # there can be a space in the unit
```
Comment (jonathankwok) [inline lib/measured/parser.rb]: `can`, or `must`? 

`+` means `must`, since it's one or more, `*` means `can` because it's zero or more.

Comment (kmcphillips) [inline lib/measured/parser.rb]: Yeah. There _can_ be a space, but then after the space there must be another thing. But this whole non-capturing group is optional. I can make that clearer.
